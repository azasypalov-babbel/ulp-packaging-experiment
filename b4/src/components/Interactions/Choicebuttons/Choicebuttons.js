import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { compose } from 'redux';
import { attemptType } from '../shared/scene/scenePropTypes';
import { Shake } from '../Wordorder/animations';


import Sheet from '../../shared/Sheet';
import { renderInBottomLayout } from '../../shared/BottomLayout';
import ChoiceItem from '../../shared/ChoiceItem/ChoiceItem';

const StyledSheet = styled(Sheet)`
  & > div {
    display: flex;
    flex-direction: column;
    & > div {
      width: 100%;
      display: flex;
      justify-content: center;
    }
    & > * + * {
      margin-top: 0.75rem;
    }
  }
`;

export const Choicebutton = ({ attempt, showFeedback, sentence, ...props }) => {
  const used = attempt?.inputText === sentence;
  const isMistaken = showFeedback && used && !attempt?.solved;
  const isCorrect = showFeedback && used && attempt?.solved;
  const ref = useRef();
  useEffect(() => {
    if (isMistaken) {
      ref?.current?.animate();
    }
  }, [ref, isMistaken]);
  return (
    <Shake ref={ref}>
      {() =>
        <ChoiceItem
          mistakeFeedback={isMistaken}
          correctFeedback={isCorrect}
          disabled={showFeedback && !used}
          {...props}
        >
          {sentence}
        </ChoiceItem>
      }
    </Shake>
  );
};

Choicebutton.propTypes = {
  sentence: PropTypes.string.isRequired,
  attempt: attemptType,
  showFeedback: PropTypes.bool
};

export const FullWidthChoicebutton = styled(Choicebutton)`
  max-width: revert;
`;

export const Choicebuttons = ({
  attempt,
  showFeedback,
  targetChoices,
  onAttempt
}) => {
  return (
    <StyledSheet onClick={(event) => event.stopPropagation()}>
      {targetChoices.map(({ sentence, correct }, index) => {
        return (
          <Choicebutton
            key={index}
            showFeedback={showFeedback}
            attempt={attempt}
            sentence={sentence}
            data-position={index}
            data-choice={sentence}
            data-correct={showFeedback ? undefined : correct}
            listenToKey={String(index + 1)}
            onClick={() => onAttempt(sentence)}
            shortcutsVisible
          >
            {sentence}
          </Choicebutton>
        );
      })}
    </StyledSheet>
  );
};

Choicebuttons.propTypes = {
  attempt: attemptType,
  showFeedback: PropTypes.bool,
  targetChoices: PropTypes.arrayOf(PropTypes.shape({
    sentence: PropTypes.string.isRequired,
    correct: PropTypes.bool.isRequired
  })).isRequired,
  onAttempt: PropTypes.func.isRequired
};

export default compose(
  renderInBottomLayout,
)(Choicebuttons);
