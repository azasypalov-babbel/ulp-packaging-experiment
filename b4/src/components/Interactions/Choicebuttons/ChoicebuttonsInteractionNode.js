import React from 'react';
import PropTypes from 'prop-types';
import { interactiveNodeType } from '../shared/scene/scenePropTypes';
import { isInteractionKey } from '../../../lib/keyboardEvents';
import ControlledGap from '../shared/Write/ControlledGap';
import Choicebuttons from './Choicebuttons';
import { INTERACTION_STATES } from '../shared/scene/useScene';

const ChoicebuttonsInteractionNode = ({
  node,
  interactionState,
  onAttempt
}) => {
  const handleKeyDown = (event) => {
    if (isInteractionKey(event, ['Shift+Escape'])) {
      event.preventDefault();
      const correctSolution = node.targetChoices.find(({ correct }) => correct);
      onAttempt(correctSolution.sentence);
      return;
    }
  };

  if (!node.active) {
    return <ControlledGap {...node} />;
  }
  return <>
    <ControlledGap
      {...node}
      onKeyDown={handleKeyDown}
      active={interactionState !== INTERACTION_STATES.INACTIVE}
      showFeedback={interactionState === INTERACTION_STATES.FEEDBACK}
    />
    {(interactionState === INTERACTION_STATES.INPUT ||
     interactionState === INTERACTION_STATES.FEEDBACK) &&
      <Choicebuttons
        showFeedback={interactionState === INTERACTION_STATES.FEEDBACK}
        attempt={node?.attempt}
        targetChoices={node.targetChoices || []}
        onAttempt={onAttempt}
      />
    }
  </>;
};

ChoicebuttonsInteractionNode.propTypes = {
  node: interactiveNodeType.isRequired,
  interactionState: PropTypes.string.isRequired,
  onAttempt: PropTypes.func.isRequired
};


export default ChoicebuttonsInteractionNode;
