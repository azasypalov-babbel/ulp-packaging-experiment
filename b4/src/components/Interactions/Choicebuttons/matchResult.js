import { RESULTS } from '../../../lib/matchingUtils/evaluate';

export const matchResult = ({ targetTexts, text }) => {
  const [correctSentence] = targetTexts.filter((target) =>  text === target);
  const solved = Boolean(correctSentence);
  const feedbackType = solved ? RESULTS.CORRECT : RESULTS.INCORRECT;

  const match = {
    text: solved ? correctSentence : targetTexts[0],
    inputText: text,
    feedbackType,
    solved
  };

  return Promise.resolve(match);
};
