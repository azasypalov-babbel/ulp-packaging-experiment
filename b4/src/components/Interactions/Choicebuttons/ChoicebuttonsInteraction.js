import React, { useCallback, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import useScene, { INTERACTION_STATES } from '../shared/scene/useScene';
import { matchResult } from './matchResult';
import useHasLatestAttempt from '../shared/Write/hasLatestAttempt';
import Text from '../../shared/Text';
import ChoicebuttonsInteractionNode from './ChoicebuttonsInteractionNode';
import useFeedbackSounds from '../shared/useFeedbackSounds';
import { getActiveNode, getItem, getMistakeCount, isGap } from '../shared/scene/selectors';
import { play, requestPrompt } from '../shared/scene/actions';

const quantiseStates = ({ active, activeNode, feedbackDismissed }) => {
  if (!active || !activeNode) return INTERACTION_STATES.INACTIVE;

  const { number, pending } = activeNode.attempt;
  if (number && !pending && !feedbackDismissed) return INTERACTION_STATES.FEEDBACK;

  return INTERACTION_STATES.INPUT;
};

export const ChoicebuttonsInteraction = ({
  item,
  active,
  learnLanguageAlpha3,
  onAttempt,
  onFinish
}) => {
  const [
    scene,
    dispatch
  ] = useScene({
    learnLanguageAlpha3,
    item,
    matchResult
  });

  const [playFeedback, isFeedbackSoundDone] = useFeedbackSounds({
    resetDependencies: [getActiveNode(scene)]
  });
  const [feedbackDismissed, setFeedbackDismissed] = useState(true);

  const { nodes } = getItem(scene);
  const activeNode = getActiveNode(scene);

  const interactionState = quantiseStates({ active, activeNode, feedbackDismissed });

  const latestAttempt = useHasLatestAttempt(activeNode && activeNode.attempt);

  const mistakeCount = getMistakeCount(scene);

  const handleAttempt = (attemptText) => {
    if (feedbackDismissed === true) {
      setFeedbackDismissed(false);
      dispatch(requestPrompt(attemptText));
    }
  };

  const handleRetry = useCallback(() => {
    setFeedbackDismissed(true);
  }, [setFeedbackDismissed]);

  const handleProceed = useCallback(() => {
    setFeedbackDismissed(true);
    dispatch(play());
  }, [setFeedbackDismissed, dispatch]);

  useEffect(() => {
    if (latestAttempt && active) {
      playFeedback(latestAttempt.solved);
      onAttempt({ attempt: latestAttempt });

      const isSolved = latestAttempt.solved;
      setTimeout(() => {
        if (isSolved) {
          handleProceed();
        } else {
          handleRetry();
        }
      }, 1000);
    }
  }, [latestAttempt, onAttempt, active, handleProceed, playFeedback, handleRetry]);

  useEffect(() => {
    if (!activeNode) {
      dispatch(play());
    }
  }, [activeNode, dispatch]);

  useEffect(() => {
    if (scene.completed && active && isFeedbackSoundDone) {
      onFinish(mistakeCount);
    }
  }, [scene.completed, onFinish, mistakeCount, isFeedbackSoundDone, active]);

  return (
    <div>
      {nodes.map((node, idx) => isGap(node)
        ? <ChoicebuttonsInteractionNode
          key={idx}
          node={node}
          learnLanguageAlpha3={learnLanguageAlpha3}
          interactionState={interactionState}
          onAttempt={handleAttempt}
          onRetry={handleRetry}
          onProceed={handleProceed}
        />
        : <Text key={idx} color="spaceGray" dangerouslySetInnerHTML={{ __html: node.text }} />
      )}
    </div>
  );
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3
});

ChoicebuttonsInteraction.propTypes = {
  active: PropTypes.bool.isRequired,
  item: PropTypes.object.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  onAttempt: PropTypes.func.isRequired,
  onFinish: PropTypes.func.isRequired
};

ChoicebuttonsInteraction.trainerConfig = {

};

export default connect(mapStateToProps)(ChoicebuttonsInteraction);
