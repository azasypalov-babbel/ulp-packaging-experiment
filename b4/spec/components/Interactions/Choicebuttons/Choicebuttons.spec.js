import React from 'react';
import { shallow } from 'enzyme';
import { Choicebuttons } from '../../../../src/components/Interactions/Choicebuttons/Choicebuttons';


const defaultProps = {
  targetChoices: [
    {
      sentence: 'k',
      correct: false
    },
    {
      sentence: 'K',
      correct: true
    }
  ],
  onAttempt: jest.fn()
};

describe('Choicebuttons', () => {
  let component;

  beforeEach(() => {
    component = shallow(<Choicebuttons {...defaultProps} />);
  });

  test('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  test('should execute callback props when clicking one of the button', () => {
    component.find('Choicebutton').first().simulate('click');

    expect(defaultProps.onAttempt).toHaveBeenCalledTimes(1);
  });
});

