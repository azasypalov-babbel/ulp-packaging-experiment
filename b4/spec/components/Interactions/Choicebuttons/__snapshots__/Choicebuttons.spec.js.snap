// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choicebuttons should render correctly 1`] = `
<Styled(Sheet)
  onClick={[Function]}
>
  <Choicebutton
    data-choice="k"
    data-correct={false}
    data-position={0}
    key="0"
    listenToKey="1"
    onClick={[Function]}
    sentence="k"
    shortcutsVisible={true}
  >
    k
  </Choicebutton>
  <Choicebutton
    data-choice="K"
    data-correct={true}
    data-position={1}
    key="1"
    listenToKey="2"
    onClick={[Function]}
    sentence="K"
    shortcutsVisible={true}
  >
    K
  </Choicebutton>
</Styled(Sheet)>
`;
